File: java-samples/Features/state-reissuance/contracts/src/main/java/net/corda/samples/statereissuance/contracts/LandTitleContract.java
package net.corda.samples.statereissuance.contracts;

import net.corda.core.contracts.Command;
import net.corda.samples.statereissuance.states.LandTitleState;
import net.corda.core.contracts.CommandData;
import net.corda.core.contracts.Contract;
import net.corda.core.transactions.LedgerTransaction;

import static net.corda.core.contracts.ContractsDSL.requireThat;

// ************
// * Contract *
// ************
public class LandTitleContract implements Contract {

    // A transaction is valid if the verify() function of the contract of all the transaction's input and output states
    // does not throw an exception.
    @Override
    public void verify(LedgerTransaction tx) {
        if(tx.getCommands().size() == 0){
            throw new IllegalArgumentException("One command Expected");
        }

        Command command = tx.getCommand(0);
        if(command.getValue() instanceof Commands.Issue)
            verifyIssue(tx);

        else if(command.getValue() instanceof Commands.Transfer)
            verifyTransfer(tx);

        else if (command.getValue() instanceof Commands.Exit)
            verifyExit(tx);
    }

    private void verifyIssue(LedgerTransaction tx){
        // Land Title Issue Contract Verification Logic goes here
        if(tx.getOutputStates().size() != 1) throw new IllegalArgumentException("One Output Expected");

        Command command = tx.getCommand(0);
        if(!(command.getSigners().contains(((LandTitleState)tx.getOutput(0)).getIssuer().getOwningKey())))
            throw new IllegalArgumentException("Issuer Signature Required");
    }

    private void verifyTransfer(LedgerTransaction tx){
        // Land Title Transfer Contract Verification Logic goes here
        Command command = tx.getCommand(0);
        LandTitleState landTitleState = (LandTitleState) tx.getInput(0);
        if (!(command.getSigners().contains(landTitleState.getIssuer().getOwningKey())) &&
                (landTitleState.getOwner() != null
                        && command.getSigners().contains(landTitleState.getOwner().getOwningKey())))
            throw new IllegalArgumentException("Issuer and Owner must Sign");
    }

    private void verifyExit(LedgerTransaction tx){
        // Land Title Exit Contract Verification Logic goes here

        if(tx.getOutputStates().size() != 0) throw new IllegalArgumentException("Zero Output Expected");
        if(tx.getInputStates().size() != 1) throw new IllegalArgumentException("One Input Expected");

        Command command = tx.getCommand(0);
        if(!(command.getSigners().contains(((LandTitleState)tx.getInput(0)).getIssuer().getOwningKey())))
            throw new IllegalArgumentException("Issuer Signature Required");
    }


    // Used to indicate the transaction's intent.
    public interface Commands extends CommandData {
        class Issue implements Commands {}
        class Transfer implements Commands {}
        class Exit implements Commands {}
        class Reissue implements Commands {}
    }
}

File: kotlin-samples/Features/state-reissuance/contracts/src/main/kotlin/net/corda/samples/statereissuance/contracts/LandTitleContract.kt
package net.corda.samples.statereissuance.contracts

import net.corda.core.contracts.CommandData
import net.corda.core.contracts.Contract
import net.corda.core.transactions.LedgerTransaction
import net.corda.samples.statereissuance.states.LandTitleState

class LandTitleContract : Contract {

    override fun verify(tx: LedgerTransaction) {
        if (tx.commands.isEmpty()) {
            throw java.lang.IllegalArgumentException("One command Expected")
        }

        val command = tx.getCommand<CommandData>(0)

        when (command.value) {
            is Commands.Issue -> {
                verifyIssue(tx)
            }
            is Commands.Transfer -> {
                verifyTransfer(tx)
            }
            is Commands.Exit -> {
                verifyExit(tx)
            }
        }
    }

    private fun verifyIssue(tx: LedgerTransaction) {
        // Land Title Issue Contract Verification Logic goes here
        if (tx.outputStates.size != 1) throw IllegalArgumentException("One Output Expected")
        val command = tx.getCommand<CommandData>(0)
        if (!command.signers.contains((tx.getOutput(0) as LandTitleState).issuer.owningKey))
            throw IllegalArgumentException("Issuer Signature Required")
    }

    private fun verifyTransfer(tx: LedgerTransaction) {
        // Land Title Transfer Contract Verification Logic goes here
        val command = tx.getCommand<CommandData>(0)
        val landTitleState = tx.getInput(0) as LandTitleState
        if (!command.signers.contains(landTitleState.issuer.owningKey) && (landTitleState.owner != null && command.signers.contains(landTitleState.owner.owningKey)))
            throw java.lang.IllegalArgumentException("Issuer and Owner must Sign")
    }

    private fun verifyExit(tx: LedgerTransaction) {
        // Land Title Exit Contract Verification Logic goes here
        if (tx.outputStates.size != 0) throw java.lang.IllegalArgumentException("Zero Output Expected")
        if (tx.inputStates.size != 1) throw java.lang.IllegalArgumentException("One Input Expected")
        val command = tx.getCommand<CommandData>(0)
        if (!command.signers.contains((tx.getInput(0) as LandTitleState).issuer.owningKey)) throw java.lang.IllegalArgumentException("Issuer Signature Required")
    }

    interface Commands : CommandData {
        class Issue : Commands
        class Transfer : Commands
        class Exit : Commands
        class Reissue : Commands
    }
}